ios/Classes/SharedStorage.swift

```diff
@@ -18,38 +19,84 @@ public enum SharedStorage: Int {
 }
 
 /// Move file at [filePath] to shared storage [destination] with optional [directory]
-public func moveToSharedStorage(filePath: String, destination: SharedStorage, directory: String) -> String? {
-    guard FileManager().fileExists(atPath: filePath)
-    else {
-        os_log("Cannot move to shared storage: file %@ does not exist", log: log, type: .error, filePath)
-        return nil
-    }
-    let fileUrl = NSURL(fileURLWithPath: filePath)
-    guard let directory = try? directoryForSharedStorage(destination: destination, directory: directory) else {
-        os_log("Cannot move to shared storage: no permission for directory %@", log: log, type: .error, directory)
-        return nil
+public func moveToSharedStorage(filePath: String, destination: SharedStorage, directory: String, complete: @escaping (String?) -> Void) {
+    switch destination {
+    case .images:
+        saveImageToPhotos(filePath: filePath, complete: complete)
+    case .video:
+        saveVideoToPhotos(filePath: filePath, complete: complete)
+    default:
+        complete(moveToFakeSharedStorage(filePath: filePath, destination: destination, directory: directory))
     }
-    if !FileManager.default.fileExists(atPath: directory.path) {
-        do
-        {
-            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories:  true)
-        } catch {
-            os_log("Cannot move to shared storage: failed to create directory %@: %@", log: log, type: .error, directory.path, error.localizedDescription)
-            return nil
+}
+
+private func saveImageToPhotos(filePath: String, complete: @escaping (String?) -> Void) {
+    var localId: String?
+
+    PHPhotoLibrary.shared().performChanges({
```

Comment:
I believe that before you `performChanges` you need to check for authorization to do so, using:
* [autohorizationStatus](https://developer.apple.com/documentation/photokit/phphotolibrary/3616052-authorizationstatus) and [requestAuthorization](https://developer.apple.com/documentation/photokit/phphotolibrary/3616053-requestauthorization) for iOS 14+
* [autohorizationStatus](https://developer.apple.com/documentation/photokit/phphotolibrary/1620745-authorizationstatus) and [requestAuthorization](https://developer.apple.com/documentation/photokit/phphotolibrary/1620736-requestauthorization) for <iOS14

The `requestAuthorization` calls require a handler that is called once the authorization has been obtained (or not), so you should move the `performChanges` block to execute in that handler, after confirming you have the appropriate `PHAuthorizationStatus`

Reply:
This is something I wasn't super sure on.  Do we want to leave this responsibility outside of this library (e.g. ask the client to use the `permission_handler` library before calling `moveToSharedStorage`)?

If you want to request authorization from _within_ this library, I think the proper move would be to refactor and handle any authorization rejections that come up, and pass those along to the calling function (on the flutter side).  Otherwise, there's no way for the calling library to know when to retry/abort/show-a-notification-about-changing-permissions-via-settings.

Reply:
Yes, that is an issue. We haev teh same issue with notification permissions, and the choice I've made there is to make the authorization request the first time it is required. You are correct that if the user declines the permission, there is no programmatic way to ask for it again (the plugin will assume that decision is permanent, and not show notifications unless the user goes to settings and manually grants the permission).

While not perfect, this is easiest and reasonable, IMHO.  If the developer wants the option to ask again, then they should use a permission management package (several of which exist), and the result of those permissions do apply to the background_downloader, so there is an option for developers that need it.

For the photos library I think we should do the same: ask for permission if needed (i.e. if not already granted or denied) and return `null` from `moveToSharedStorage` if the permission is denied (and log something to that effect).  Importantly, this is also consistent with how it works on Android.

Reply:
First: sure, this is your library, and it's up to you.

But: As soon as a consumer of this library has the intention of putting this into production, then the decline flow _needs_ to be handled.  The user experience is brutal otherwise.  And this just returns `null`.  No message, and nothing actionable.

I know, because I've gone through the swift code, that this is the case.  But this is opaque to the library consumer.

IMO better to have zero permissions handling than half-baked permissions handling.

Reply:
> IMO better to have zero permissions handling than half-baked permissions handling.

Because it makes it clear upfront to the consumer of the library what is needed.

Reply:
Re-reading this, it came off a lot firmer than I intended.  I do understand that you have already established a pattern of how `background_downloader` handles permissions, and I'm happy to follow it.  I just wanted to share some of my thoughts about potential pitfalls.

Reply:
I understand what you're saying. One option is to add methods to the FileDownloader to manage permissions directly. For example, you could use one method to get the status of a certain permission, and another to ask for permission.  It gets a bit tricky in that it needs to work for both iOS and Android, and of course the permissions aren't exactly the same. 

I'm hesitant because I worry I'll end up re-implementing what permission plugins already do well today. My question to you therefore is: if you need to manage a user declining certain permissions, then why don't you use one of those permission plugins? And if you don't need to manage the decline (e.g. I don't think you need to manage a user declining your download notifications) then the background_downloader plugin does everything you need.  Isn't that a reasonable approach? I am genuinely trying to find the right approach here, so want to understand your perspective better.